docs/design/backups.md

```diff
@@ -0,0 +1,83 @@
+# Backup & Restore
+
+In particularly disastrous situations, we would like to be able to revert the etcd cluster to a previously known healthy state.
+This document lays out the design of how we want the healthy state to be captured, securely stored & safely restored to operator-managed etcd clusters.
+
+## Backup Resource
+
+We will introduce a new custom resource: `EtcdBackup`.
+
+```yaml
+apiVersion: etcd.improbable.io/v1alpha1
+kind: EtcdBackup
+metadata:
+  name: my-cluster-backup
+  namespace: default
+spec:
+  type: snapshot
+  destination:
+    gcsBucket:
+      bucketName: my-cluster-backup-bucket
+      creds:
+        secretRef: 
+          name: my-cluster-backup-bucket-credentials
+          namespace: default
+```
+
+This resource specifies that we desire a backup to be immediately taken as a [snapshot](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/recovery.md#snapshotting-the-keyspace) from the etcd API.
+This snapshot will be saved to a GCS bucket, accessed with credentials read from a given secret.
+The file would be named with a timestamp when the backup was started, for uniqueness.
+Backups will be taken by the controller, rather than a Job or Pod acting on behalf of the controller.
+
+Backup types could include
+  * `snapshot`: a `snapshot.db` file taken from the etcd API
+  * `keys`: a zip file containing all key-values as files
+  * `keys-json` a json file containing all key-values as a JSON file
+
+Only `snapshot` will be initially implemented.
+
+Destinations culd include
+  * `socket`: data would be sent to a TCP socket
```

Comment:
Is this a standardised thing? This one does not look like the others in terms of the semantics/API it presents

Reply:
its standard in that it represents a way of "piping" data from one app to another, a standard OS feature. i've seen it done in other apps for telemetry/log data, but i don't see a reason it wouldn't also work for binaries. might be useful if someone wants to build an integration with some proprietary storage location which we may not want to support in the OS repo.

Reply:
:+1: seems reasonable, I'm just conscious we don't have a way to transfer metadata such as the backup timestamp, or name of the cluster being backed up.

Reply:
fair concern. we can send anything we want over a socket as long as we define how its encoded - but that's a detail we can squash later as we won't be implanting this for now